modules/gdprEnforcement.js

```diff
@@ -0,0 +1,181 @@
+  if (bidderCode) {
+    const bidder = adapterManager.getBidAdapter(bidderCode);
+    gvlid = bidder.getSpec().gvlid;
+  } else {
+    utils.logWarn('Current bidder not found');
+  }
+  return gvlid;
+}
+
+/**
+ * This function takes in rules and consentData as input and validates against the consentData provided. If it returns true Prebid will allow the next call else it will log a warning
+ * @param {Object} rules enforcement rules set in config
+ * @param {Object} consentData gdpr consent data
+ * @returns {boolean}
+ */
+function validateRules(rule, consentData, currentModule, gvlid) {
+  let isAllowed = false;
+  if (rule.enforcePurpose && rule.enforceVendor) {
+    if (includes(rule.vendorExceptions, currentModule) ||
+        (utils.deepAccess(consentData, 'vendorData.purpose.consents.1') === true && utils.deepAccess(consentData, `vendorData.vendor.consents.${gvlid}`) === true)) {
+      isAllowed = true;
+    }
+  } else if (rule.enforcePurpose === false && rule.enforceVendor === true) {
+    if (includes(rule.vendorExceptions, currentModule) ||
+        (utils.deepAccess(consentData, `vendorData.vendor.consents.${gvlid}`) === true)) {
+      isAllowed = true;
+    }
+  } else if (rule.enforcePurpose === false && rule.enforceVendor === false) {
+    if ((includes(rule.vendorExceptions, currentModule) &&
+        (utils.deepAccess(consentData, 'vendorData.purpose.consents.1') === true && utils.deepAccess(consentData, `vendorData.vendor.consents${gvlid}`) === true)) ||
+        !includes(rule.vendorExceptions, currentModule)) {
+      isAllowed = true;
+    }
+  }
+  return isAllowed;
+}
+
+/**
+ * This hook checks whether module has permission to access device or not. Device access include cookie and local storage
+ * @param {Function} fn reference to original function (used by hook logic)
+ * @param {Number=} gvlid gvlid of the module
+ * @param {string=} moduleName name of the module
+ */
+export function deviceAccessHook(fn, gvlid, moduleName) {
+  let result = {
+    hasEnforcementHook: true
+  }
+  if (!hasDeviceAccess()) {
+    utils.logWarn('Device access is disabled by Publisher');
+    result.valid = false;
+    return fn.call(this, result);
+  } else {
+    const consentData = gdprDataHandler.getConsentData();
+    if (consentData && consentData.gdprApplies && consentData.apiVersion === 2) {
+      if (!gvlid) {
+        gvlid = getGvlid();
+      }
+      const curModule = moduleName || config.getCurrentBidder();
+      const purpose1Rule = find(enforcementRules, hasPurpose1);
+      let isAllowed = validateRules(purpose1Rule, consentData, curModule, gvlid);
+      if (isAllowed) {
+        result.valid = true;
+        return fn.call(this, result);
+      } else {
+        utils.logWarn(`User denied Permission for Device access for ${curModule}`);
+        result.valid = false;
+        return fn.call(this, result);
+      }
+    } else {
+      utils.logInfo('GDPR enforcement only applies to CMP version 2');
```

Comment:
In trying to be sensitive to overall package size, I'd like to encourage concise warning messages. Could we cut some of these down? e.g. "Enforcing TCF2 only" is less than half the length of this message.

Reply:
Updated the message